File: Data Structures & Algorithms/is-anagram/submission-3.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        // Simply create two hashset
        unordered_map< char, int> s_freq;
        unordered_map< char , int> t_freq;

        //make for s
        int sn = s.size();
        int tn = t.size();

        for (int i = 0; i < sn; i++) {
            char curr = s[i];

            if (!s_freq.count(curr)) {
                s_freq[curr] = 1;
            } else {
                s_freq[curr]++;
            }
        }

        for (int i = 0; i < tn; i++) {
            char curr = t[i];
            if (!t_freq.count(curr)) {
                t_freq[curr] = 1;
            } else {
                t_freq[curr]++;
            }
        }

        int k1 = s_freq.size();
        int k2 = t_freq.size();

        if (k1 != k2) { return false; }
        
        //now check if they are the same or not
        for (const auto& pair: s_freq) {
            char key = pair.first;
            if (!t_freq.count(key)) { return false; }
            else {
                if (s_freq[key] != t_freq[key]) { return false; }
            }
        }
        return true;
    }
};

File: Data Structures & Algorithms/palindromic-substrings/submission-3.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        
        """
        #brute force <- get all substrs and check each one of them if they are a palindrome or not

        all_substr= []

        for i in range(len(s)):
            for j in range(i,len(s)):
                all_substr.append(s[i:j+1])

        def _checkPalin(text):
            i,j = 0, len(text) - 1

            while i <= j:
                if text[i] != text[j]:
                    return False
                i += 1
                j -= 1

            return True

        count = 0
        for ss in all_substr:
            if _checkPalin(ss):
                count += 1
        
        return count
        """

        #same logic as the prev one
        total = 0

        #lets check odd length first
        for curr in range(len(s)):
            i,j = curr - 1, curr + 1
            print(i,j)

            while i >= 0 and j < len(s) and s[i] == s[j]:
                total += 1
                i -= 1
                j += 1

        
        #lets check even length now
        for curr in range(len(s)):
            l,r = curr - 1, curr
            
            while l >= 0 and r < len(s) and s[l] == s[r]:
                total += 1
                l -= 1
                r += 1

        return total + len(s)